docs/src/_contribute/cli.md

```diff
@@ -0,0 +1,255 @@
+and is built with the [Python package: click](https://click.palletsprojects.com/en/8.1.x/).
+
+## Getting Set Up
+
+See the prerequisite section for instructions on how to set up a development environment and dependencies.
+
+## CLI Design Guidelines
+
+The `meltano` CLI is the primary interface for interacting with Meltano. It's often the first introduction to meltano
+that a users experiences when browsing the docs or experimenting with the CLI. As such, we aim to provide a clear and
+consistent interface that is easy to use. This guide aims to codify the design guidelines for the CLI so that
+contributors can easily update or extend the CLI with new features, commands, options, and enhancements.
+
+This document is somewhat aspirational and a work in progress. Portions of the CLI may violate the design and style
+guidelines, but we aim to update the CLI overtime to ensure that it is as consistent as possible.
+
+### Consistent definition of Groups, Commands, Sub-commands, Arguments, and Options
+
+Groups, Commands, Sub-commands, Arguments, Options, and Flags often have slightly different meanings for users. So are
+defined as follows for use in `meltano`.
+
+- Group is a group of commands that are related to a specific area of the CLI, typically a specific meltano feature.
+  For example, the `meltano schedule` feature-group contains commands related to managing Meltano schedules.
+- command/sub_command are used to perform specific tasks or used to group a set of command around a specific feature
+  or task. Sub-commands take the form of parameters, but require additional arguments or options themselves.
+- Arguments are positional parameters that are passed to a command. Arguments do not require additional options or
+  arguments - otherwise they would be considered a sub-command.
+- Options (switch, option flags, or flags) are options that alter the behavior (e.g. `--dry-run/--verbose`) or named
+  input options (e.g. `--tasks=`). They come in two forms. A long form: `--option-name`, and a short form: `-o`.
+
+Given a command like `meltano job set JOB_NAME --tasks=[<task>...]`:
+
+- `meltano` is the global click command group for Meltano. This is the main entry point to the CLI. When we refer to global options this is the level we refer to.
+- `job` is the click command group for the `job` command. This is the main command group for the `job` command. This is typically the top level for a Meltano feature.
+- `set` is a `sub-command` of the `job` command.
+- `JOB_NAME` is an `argument` of the `set` sub-command.
+- `--tasks` is named `option` of the `set` sub-command.
+
+For a technical explanation of how commands and groups work see https://click.palletsprojects.com/en/8.0.x/commands/
+
+### Use of short and long options
+
+The options you expect to be used most frequently should always include both the long AND short variation of the option.
+For example `--help` and `-h`, `--tasks` and `-t`, `--force` and `-f`.
```

Comment:
I might have a less favorable opinion about short options (and also abbreviations) in general, since I find that the benefit in brevity comes at a high cost in readability and having to frequently revisit the docs and/or `--help` output. (Remembering that `-e` means one thing but `-E` means something different, and then `-o` means something whose name _also_ starts with `e` but we had already used `-e` so the short option had to be something completely different.)

I would be happy to remove both of these two topics (short options and abbreviations) from this iteration of the CLI guidelines doc, or else perhaps mitigate using one or more of the below.

1. We can suggest that short options and abbreviations should be discussed and agreed to in the spec ahead of time, prior to development starting. (Prevents stalling a PR later for discussion which otherwise would have been ready to go.)
2. We can suggest that short options and abbreviations be _excluded_ from initial feature releases and instead added after the fact in a follow-on PR. (Similar mitigation as above. We just treat abbreviations as QOL improvements after the fact, and therefor by definition they won't be blocking critical functionality and we won't rush to decisions without proper consideration.)
3. We can just say we don't prefer short options or abbreviations at this time, but that we plan to add them in a considered manner later in the product lifecycle.


Any or all of the above recommendations would prevent us from blocking new features from going out due to internal deliberations of acceptable convenience methods. 

@tayloramurphy and @pandemicsyn - Would it be okay with you both if we take a conservative approach to these for the time being? As the adage goes, "easier to add later than to try to remove after adding".

Also, assuming we keep some iteration of these topics in this doc ("short options" and "abbreviations"), and since they are closely related, what do you think about combining them into a single section, perhaps with subheaders? Then we can offer some type of guidance of when to use them and how to propose new ones as a single guidance for both topics.

Wdyt?

Reply:
@aaronsteers I don't have strong opinions on short flags and am ok with a more conservative approach 👍

Reply:
> Personally, I find that the benefit in brevity comes at a high cost in readability and having to frequently revisit the docs and/or --help output. 

I think that's only true until you're comfortable with the application. And having purely long options doesn't mean you won't visit help as often - and to be frank I think the opposite is true. I have a much better chance of remember that I need `grep -b` than rembering that I need `grep --before-context`. My push back is, day to day - how often do you folks type `ls --long` vs `ls -l` , `grep -c` vs `grep --count`, `grep -b vs grep --before-context`?

> @tayloramurphy and @pandemicsyn - Would it be okay with you both if we take a conservative approach to these for the time being? As the adage goes, "easier to add later than to try to remove after adding".

I think saying "flags should be in the spec and short flags should be vetted" is totally acceptable. But why completely remove the guidance - that just opens us up for more bike shedding, because it means there's not even standard to propose or a standard for a proposed flag to adhere too.

Reply:
@tayloramurphy I'll add that it makes intuitive sense to me, and will likely result in a more enjoyable UX for more experienced users, to have the short flags available but I understand the bikeshedding concerns. A considered approach feels reasonable but I would not be in favor a blanket "no short flags".

Reply:
I agree with @pandemicsyn. For experienced users, it's easier to remember the short form and I can't think of a CLI that I used heavily where I don't prefer the short flag: `nox -rs`, `poetry install -E docs`, etc.

Now, we don't have to add a short form to _all_ options, just the most used ones. We should leave the other open and offer clear guidance for the team and contributors.

Some middle would be to require long flags be used in documented examples, tutorials, etc. so it's more clear what those flags do, and where possible linking to the full command documentation, where short flags are shown alongside their long counterparts.

Reply:
> Now, we don't have to add a short form to all options, just the most used ones. We should leave the other open and offer clear guidance for the team and contributors.

+1 , right now the guidelines say `The options you expect to be used most frequently should always include both the long AND short variation of the option.` 

 Could definitely revise that to say something along the lines `reserve short flag usage only for what you expect the most frequently used options to be, if you feel the short flag can be confused with another, bias towards using the long variant only`

Reply:
> > Now, we don't have to add a short form to all options, just the most used ones. We should leave the other open and offer clear guidance for the team and contributors.
> 
> +1 , right now the guidelines say `The options you expect to be used most frequently should always include both the long AND short variation of the option.` 
> 
>  Could definitely revise that to say something along the lines `reserve short flag usage only for what you expect the most frequently used options to be, if you feel the short flag can be confused with another, bias towards using the long variant only`

Yeah, this makes a lot of sense to me. My point in the first comment above is not that we shouldn't have short flags but only that I want to be mindful ahead of time where collisions are going to occur. And figuring them out before the feature is under development, or after the feature had shipped, is preferable to moving too fast or too slow when the PRs are otherwise ready to ship.

Reply:
To come to a short flag strategy we can all feel good about, I do think it'd be helpful to take a pass over existing already shipped features. We'd have some good lessons learned there and would likely come out of that exercise with a pretty holistic set of leanings based on existing use cases and usage patterns.

(I really liked, for example, the uppercase -E for environments because it's a global setting and the lowercase -e being available on specific commands.)

Reply:
> I do think it'd be helpful to take a pass over existing already shipped features.

@aaronsteers that definitely makes sense. I was planning on pop'ing a follow-up issue to clean up some of our help output (to conform to our new guidelines), and propose some flag additions. Would you prefer I send these along as part of this PR?

Reply:
@aaronsteers just wanted to give this thread bump, to keep this PR rolling.

Reply:
@aaronsteers @tayloramurphy I updated the wording to more strongly indicate that short flags should be used only when unambiguous, and the long version is excessively long or expected to be frequently used. I also switched the tone to from "you should" to "you may".

I also clarified how they should be presented in docs and that we'd prefer that only long options be used in things like getting start guides and walkthroughs. 

I also completely nuked the abbreviations/aliases section. 

Are these changes enough to keep this rolling?

Reply:
@pandemicsyn love it. this already has my stamp of approval ✅